lib/buffer.js

```diff
@@ -176,12 +176,25 @@ Buffer.from = function(value, encodingOrOffset, length) {
   if (isAnyArrayBuffer(value))
     return fromArrayBuffer(value, encodingOrOffset, length);
 
+  if (value == null)
+    throw new TypeError(kFromErrorMsg);
+
+  if (typeof value === 'number')
+    throw new TypeError('"value" argument must not be a number');
+
+  const valueOf = value.valueOf();
+  if (valueOf !== value)
+    return Buffer.from(valueOf, encodingOrOffset, length);
+
   var b = fromObject(value);
```

Comment:
Now that we're supposedly no longer bound by Crankshaft's inlining limitations, perhaps we might also move the `isUint8Array()` case from `fromObject()` into this function, right below the `isAnyArrayBuffer()` case.